Microsoft/checkedc-clang
# Bounds for call expressions where the bounds depend on call results.

This change implements inference of bounds for call expressions for cases where the bounds depend on the result of the call expressions.   This changes adds inference for `count` and `byte_count` return bounds, including cases where a bounds-safe interface adds these as return bounds.  An additional case that is not yet implemented is when the return bounds uses the `_Return_value` expression.

We didn't implement the inference before because we needed a way to represent the result of evaluating an expression.  In PR #561, we extended the clang IR with temporary variables.    A temporary variable can be bound to the result of an expression and then used later within the enclosing top-level expression.   This change documents the temporary variable extension to the IR made in PR #561.

This change adds temporary variables for call expressions where the bounds depends on the results of the call..   It moves inference for bounds for call expressions to a separate function that also takes the temporary variable (if any) that will hold the result of the call.   It fills in cases for count and byte_count.   It extends the lowering from clang IR to LLVM IR to handle the case where a scalar expression result is bound to a temporary variable.

Before this change, we were returning Bounds(any) for call expressions where we could not infer bounds.  This disabled bounds checking for the result of a call (for example, `f(len)[]5]` would not have bounds checking. It also allowed static checking of bounds declarations where the calls was the right-hand side of assignment or initializer to pass.

This change adds equality facts for initializers (T lhs = rhs) and assignment (lhs = rhs) where the right-hand side binds a temporary variable v.  We add the fact that lhs == v. 

With this change, the compiler now issues additional warnings about being unable to prove the validity of bounds.  This happens for some benchmarks in PtrDst and Olden.  The compiler doesn't understand when some pointer arithmetic expressions are equivalent.   I open issue #596 to track this.

I updated the Checked C clang rewriter to handle additional places where temporary variables can appear.

Testing:
- Update existing AST tests for bounds inference for calls to have the inferred bounds.
- Add new tests of checking bounds declarations to the Checked C clang repo.
- Update Checked C repo tests for cases where new warning messages appear.  Tagged them with GitHub issues.
- Passing existing Checked C, Checked C clang, and LLVM test suite suites.
